### PR TITLE
Support `mvzip` eval function

### PR DESCRIPTION
**Pull request by ahkcs**
## **Description**

The `mvzip` function combines values from two multivalue fields pairwise with a delimiter.
It stitches together corresponding elements from each field, similar to Python’s `zip()` function.

The function supports two modes of operation:

1. **Default delimiter:**

   ```ppl
   mvzip(mv_left, mv_right)
   ```

   Combines fields using a comma (`,`) as the default delimiter.

2. **Custom delimiter:**

   ```ppl
   mvzip(mv_left, mv_right, delimiter)
   ```

   Combines fields using the specified delimiter.

---

### **Key Features**

* **Pairwise combination:** Combines 1st element of left with 1st of right, 2nd with 2nd, etc.
* **Stops at shorter length:** Processing stops at the length of the shorter field (Python `zip()` behavior).
* **Scalar handling:** Treats scalar values as single-element arrays.
* **Null handling:** Returns `null` if either input is `null`.
* **Default delimiter:** Uses comma (`,`) when delimiter is not specified.

---

### **Basic Usage with Default Delimiter**

```ppl
source=people 
| eval hosts = array('host1', 'host2'), ports = array(80, 443), nserver = mvzip(hosts, ports) 
| fields nserver
# Returns: [host1,80, host2,443]

source=accounts 
| eval result = mvzip(firstname, lastname) 
| fields result
# Returns: [Amber,Duke]
```

---

### **Custom Delimiter**

```ppl
source=people 
| eval arr1 = array('a', 'b', 'c'), arr2 = array('x', 'y', 'z'), result = mvzip(arr1, arr2, '|') 
| fields result
# Returns: [a|x, b|y, c|z]

source=accounts 
| eval result = mvzip(firstname, lastname, ' ') 
| fields result
# Returns: [Amber Duke]
```

---

### **Different Length Arrays**

```ppl
source=people 
| eval arr1 = array(1, 2, 3), arr2 = array('a', 'b'), result = mvzip(arr1, arr2) 
| fields result
# Returns: [1,a, 2,b]
# Note: Stops at length of shorter array
```

---

### **Nested mvzip Calls**

```ppl
source=people 
| eval field1 = array('a', 'b'), field2 = array('c', 'd'), field3 = array('e', 'f'), result = mvzip(mvzip(field1, field2, '|'), field3, '|') 
| fields result
# Returns: [a|c|e, b|d|f]
```

---

### **Null Handling**

```ppl
source=people 
| eval result = mvzip(nullif(1, 1), array('test')) 
| fields result
# Returns: null
```

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
 - [x] New functionality has javadoc added.
 - [ ] New functionality has a user manual doc added.
- [ ] New PPL command [checklist](https://github.com/opensearch-project/sql/blob/main/docs/dev/ppl-commands.md) all confirmed.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff` or `-s`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ahkcs** [comment]
> QQ: can SQL's `ARRAYS_ZIP` be used for this?

Thanks for the question! I considered ARRAYS_ZIP but it's not suitable for mvzip due to semantic differences:
  1. Return Type:
    - ARRAYS_ZIP → ARRAY<STRUCT> (e.g., [STRUCT(0:'a', 1:'x'), STRUCT(0:'b', 1:'y')])
    - mvzip → ARRAY<VARCHAR> (e.g., ['a|x', 'b|y'])
  2. Delimiter Requirement:
    - mvzip requires custom delimiter support for string concatenation
    - ARRAYS_ZIP creates structured data without string formatting

File: docs/user/ppl/functions/collection.rst [hunk 1]
```diff
@@ -356,3 +356,45 @@ Example::
...
+
+    os> source=people | eval arr1 = array('a', 'b', 'c'), arr2 = array('x', 'y', 'z'), result = mvzip(arr1, arr2, '|') | fields result | head 1
+    fetched rows / total rows = 1/1
+    +---------------+
+    | result        |
+    |---------------|
+    | [a|x,b|y,c|z] |
+    +---------------+
+
+    os> source=people | eval arr1 = array(1, 2, 3), arr2 = array('a', 'b'), result = mvzip(arr1, arr2) | fields result | head 1
```
Comment (dai-chen) [inline docs/user/ppl/functions/collection.rst]: Add or modify the example for nested `mvzip` with 3 multi values?

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      // mvzip returns an array of VARCHAR (strings)
+      RelDataType elementType = typeFactory.createSqlType(SqlTypeName.VARCHAR);
+      return createArrayType(
+          typeFactory, typeFactory.createTypeWithNullability(elementType, true), true);
+    };
+  }
+
+  @Override
+  public UDFOperandMetadata getOperandMetadata() {
+    return null;
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java]: The first 2 argument type must be multiarray/array? If so, no need to accept object in implemention method `mvzip`?

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipCore.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    for (int i = 0; i < minLength; i++) {
+      Object leftValue = leftList.get(i);
+      Object rightValue = rightList.get(i);
+
+      // Combine the values with the delimiter
+      String combined = leftValue + delimiter + rightValue;
+      result.add(combined);
+    }
+
+    return result.isEmpty() ? null : result;
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipCore.java]: So if one or both array is empty, the result is null?

File: docs/user/ppl/functions/collection.rst (hunk 1 above)
Comment (ahkcs) [reply docs/user/ppl/functions/collection.rst]: Added example for nested `mvzip` with 3 arrays

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java (hunk 2 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java]: Updated the `getOperandMetadata` method to only accept Array for the first 2 arguments

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipCore.java (hunk 3 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipCore.java]: Updated the NullPolicy to be NullPolicy.ANY, also updated the IT tests and UT for cases where one or both array is empty

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+   * (rather than List) because Calcite's reflection-based method lookup uses Object.class. Type
+   * validation is enforced at compile time via {@link #getOperandMetadata()}.
+   *
+   * @param left The left multivalue array
+   * @param right The right multivalue array
+   * @param delimiter The delimiter to use for joining values (default: ",")
+   * @return A list of combined values, empty list if either array is empty, or null if either input
+   *     is null
+   */
+  public static List<Object> mvzip(Object left, Object right, String delimiter) {
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java]: Although string array argument type, Calcite requires `Object` arg here?

File: docs/user/ppl/functions/collection.rst [hunk 5]
```diff
@@ -394,3 +394,53 @@ Example::
...
+
+The delimiter is optional. When specified, it must be enclosed in quotation marks. The default delimiter is a comma ( , ).
+
+Argument type: mv_left: ANY, mv_right: ANY, delim: STRING (optional)
+
+Return type: ARRAY
+
+Example::
+
+    os> source=people | eval hosts = array('host1', 'host2'), ports = array(80, 443), nserver = mvzip(hosts, ports) | fields nserver | head 1
```
Comment (dai-chen) [inline docs/user/ppl/functions/collection.rst]: np: `head 1` is because `source=people` is required? Do we have an issue for adding `makeresults` command which would simplify this a lot?

**Comment by dai-chen** [comment]
> > QQ: can SQL's `ARRAYS_ZIP` be used for this?
> 
> Thanks for the question! I considered ARRAYS_ZIP but it's not suitable for mvzip due to semantic differences:
> 
> 1. Return Type:
>    - ARRAYS_ZIP → ARRAY (e.g., [STRUCT(0:'a', 1:'x'), STRUCT(0:'b', 1:'y')])
>    - mvzip → ARRAY (e.g., ['a|x', 'b|y'])
> 2. Delimiter Requirement:
>    - mvzip requires custom delimiter support for string concatenation
>    - ARRAYS_ZIP creates structured data without string formatting

Does the `mvmap` function you're working help here?

**Comment by ahkcs** [comment]
> Does the `mvmap` function you're working help here?

For `mvmap` function, we are using the existing `Transform` eval function for its implementation
https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/functions/collection.rst#transform

https://github.com/opensearch-project/sql/pull/4856

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java (hunk 4 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java]: Hi Chen, Calcite's reflection-based method lookup requires Object types - When Types.lookupMethod is called (lines 82-83, 90-91), it looks up the method signature using Object.class:
```
  Types.lookupMethod(
      MVZipFunctionImpl.class, "mvzip", Object.class, Object.class, String.class)
```

File: docs/user/ppl/functions/collection.rst (hunk 5 above)
Comment (ahkcs) [reply docs/user/ppl/functions/collection.rst]: Yes, head 1 is needed because source=people returns multiple rows. We have issue #3629 for adding a data generation command that is still open

File: core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java (hunk 4 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVZipFunctionImpl.java]: Updated to use `List.class` instead

**Comment by ykmr1224** [comment]
Can you resolve conflict, and address comment from CodeRabbit.

File: docs/user/ppl/functions/collection.rst [hunk 6]
```diff
@@ -394,3 +394,53 @@ Example::
...
+    | [1,a,2,b] |
+    +-----------+
+
+    os> source=people | eval arr1 = array(1, 2, 3), arr2 = array('a', 'b', 'c'), arr3 = array('x', 'y', 'z'), result = mvzip(mvzip(arr1, arr2), arr3) | fields result | head 1
+    fetched rows / total rows = 1/1
+    +---------------------+
+    | result              |
+    |---------------------|
+    | [1,a,x,2,b,y,3,c,z] |
+    +---------------------+
```
[the remark this answers is not in the document]
Comment (ahkcs) [reply docs/user/ppl/functions/collection.rst]: Updated

**Review by dai-chen — APPROVED** [review]
